core/contractsdk/cpp/example/xuper_relayer.cc

```diff
@@ -0,0 +1,478 @@
+    if (anchorBlockHeaderStr.size()==0) {
+        ctx->error("missing blockHeader");
+        return;
+    }
+    bool succ = anchorBlockHeader->ParseFromString(anchorBlockHeaderStr);
+    if (!succ) {
+        ctx->error("parse from string error");
+        return;
+    }
+
+    std::string blockidBuf = anchorBlockHeader->blockid();
+    std::string visualBlockid = std::string(64, '0');
+    if (!encode(blockidBuf, visualBlockid)) {
+        ctx->error("encode blockid failed");
+        return;
+    }
+    std::string metaStr;
+    const std::string metaKey = std::string(ledgerMetaBucket);
+    const std::string anchorBlockHeaderKey = std::string(blockHeaderBucket) + visualBlockid;
+    meta->set_root_blockid(visualBlockid);
+    meta->set_tip_blockid(visualBlockid);
+    meta->set_trunk_height(anchorBlockHeader->height());
+    meta->SerializeToString(&metaStr);
+    if (!ctx->put_object(anchorBlockHeaderKey, anchorBlockHeaderStr) ||
+        !ctx->put_object(metaKey, metaStr)) {
+        ctx->error("put anchorBlockHeader or ledgerMeta failed");
+        return;    
+    }
+
+    ctx->ok("initialize succeed");
+}
+
+DEFINE_METHOD(XuperRelayer, putBlockHeader) {
+    xchain::Context* ctx = self.context();
+    std::unique_ptr<relayer::LedgerMeta> meta(new relayer::LedgerMeta);
+    std::unique_ptr<relayer::InternalBlock> blockHeader(new relayer::InternalBlock);
+    // 提取ledgerMeta
+    std::string metaStr;
+    const std::string metaKey = std::string(ledgerMetaBucket);
+    if (!ctx->get_object(metaKey, &metaStr)) {
+        ctx->error("missing ledger meta info");
+        return;
+    }
+    meta->ParseFromString(metaStr);
+    // 提取blockHeader
+    std::string blockHeaderStr = ctx->arg("blockHeader");
+    bool succ = blockHeader->ParseFromString(blockHeaderStr);
+    if (!succ) {
+        ctx->error("parse from string error");
+        return;
+    }
+    std::string blockidBuf = blockHeader->blockid();
+    std::string visualBlockid = std::string(64, 'o');
+    if (!encode(blockidBuf, visualBlockid)) {
+        ctx->error("encode blockid failed");
+        return;
+    }
+    const std::string blockHeaderKey = std::string(blockHeaderBucket) + visualBlockid;
+    std::string tmp;
+    if (ctx->get_object(blockHeaderKey, &tmp)) {
+        ctx->error(visualBlockid + " has existed already");
+        return;
+    }
+    // 判断区块类型
+    std::string preHashBuf = blockHeader->pre_hash();
+    // root链
+    if (preHashBuf.size() == 0) {
+        ctx->error("preHash shouldn't be empty");
+        return;
+    }
+    std::string visualPreHash = std::string(64, 'o');
+    if (!encode(preHashBuf, visualPreHash)) {
+        ctx->error("encode preHash failed");
+        return;
+    }
+    // 判断preBlockcHeader是否存在
+    const std::string preBlockHeaderKey = std::string(blockHeaderBucket) + visualPreHash;
+    std::string preBlockHeaderStr;
+    std::unique_ptr<relayer::InternalBlock> preBlockHeader(new relayer::InternalBlock);
```

Comment:
need to parse preBlockHeader and update the preBlockHeader value?

Reply:
Yes, it's necessary. We should verify it by contract self.